File: bot.py
import logging
import os
from dotenv import load_dotenv
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Application, CommandHandler, CallbackQueryHandler, MessageHandler, filters, ContextTypes
from database import (
    add_user_channel, remove_user_channel, get_user_channels,
    add_all_channel, get_all_channels, get_channel_owner,
    add_post_request, get_pending_requests, update_post_request_status
)

# Загрузка переменных окружения из файла .env
load_dotenv()

# Получаем токен из переменных окружения
TOKEN = os.getenv('TELEGRAM_BOT_TOKEN')

if not TOKEN:
    raise ValueError("Токен не найден в переменных окружения")

# Задаем уровень логирования
logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO)
logger = logging.getLogger(__name__)

# Глобальная переменная для хранения запросов
post_requests = {}

# Функция для старта бота
async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await update.message.reply_text('Привет! Я бот для управления постами в каналах.')

# Функция для обработки команды /stats
async def stats(update: Update, context: ContextTypes.DEFAULT_TYPE):
    user_id = update.message.from_user.id
    user_channels = get_user_channels(user_id)

    if not user_channels['channels']:
        await update.message.reply_text(
            'У вас нет привязанных каналов. Используйте /addchannel @channel_name для добавления канала.'
        )
        return

    response = f'Личный кабинет пользователя {user_id}:\nПривязанные каналы:\n'
    for index, channel in enumerate(user_channels['channels'], start=1):
        response += f'{index}. @{channel}\n'
    
    response += '\nДобавить канал: /addchannel @channel_name\n'
    response += 'Удалить канал: /removechannel @channel_name\n'
    response += 'Создать запрос на взаимный пост: /createpost\n'

    await update.message.reply_text(response)

# Функция для добавления канала
async def add_channel(update: Update, context: ContextTypes.DEFAULT_TYPE):
    user_id = update.message.from_user.id
    if len(context.args) == 0:
        await update.message.reply_text('Пожалуйста, укажите название канала после команды. Пример: /addchannel @channel_name')
        return
    
    channel_name = context.args[0].replace('@', '')
    chat_id = f'@{channel_name}'

    if await check_if_bot_can_post_messages(chat_id):
        add_user_channel(user_id, channel_name)
        add_all_channel(channel_name, user_id)
        await update.message.reply_text(f'Канал @{channel_name} добавлен к вашему аккаунту.')
    else:
        await update.message.reply_text(f'Пожалуйста, убедитесь, что я имею право "Отправлять сообщения" в канале @{channel_name}.')

# Функция для удаления канала
async def remove_channel(update: Update, context: ContextTypes.DEFAULT_TYPE):
    user_id = update.message.from_user.id
    if len(context.args) == 0:
        await update.message.reply_text('Пожалуйста, укажите название канала после команды. Пример: /removechannel @channel_name')
        return
    
    channel_name = context.args[0].replace('@', '')
    remove_user_channel(user_id, channel_name)
    await update.message.reply_text(f'Канал @{channel_name} удален из вашего аккаунта.')

# Функция для проверки прав на отправку сообщений
async def check_if_bot_can_post_messages(chat_id: str) -> bool:
    try:
        chat_member = await application.bot.get_chat_member(chat_id, application.bot.id)
        return chat_member.status in ['administrator', 'creator'] and chat_member.can_post_messages
    except Exception as e:
        logger.error(f'Ошибка при проверке прав: {e}')
        return False

# Функция для создания запроса на взаимный пост
async def create_post(update: Update, context: ContextTypes.DEFAULT_TYPE):
    user_id = update.message.from_user.id
    user_channels = get_user_channels(user_id)

    if not user_channels['channels']:
        await update.message.reply_text('У вас нет привязанных каналов. Используйте /addchannel @channel_name для добавления канала.')
        return

    all_channels = get_all_channels()

    if not all_channels:
        await update.message.reply_text('В базе данных нет доступных каналов для взаимного поста.')
        return

    keyboard = [[InlineKeyboardButton(f'@{channel}', callback_data=f'create_request_{channel}')] for channel in all_channels]
    reply_markup = InlineKeyboardMarkup(keyboard)
    await update.message.reply_text('Выберите канал для взаимного поста:', reply_markup=reply_markup)

# Функция для обработки нажатия кнопок выбора канала
async def button(update: Update, context: ContextTypes.DEFAULT_TYPE):
    query = update.callback_query
    data = query.data
    user_id = query.from_user.id

    if data.startswith('create_request_'):
        channel_name = data[len('create_request_'):]
        post_requests[user_id] = channel_name

        owner_id = get_channel_owner(channel_name)
        if owner_id:
            post_template = "Пример шаблона поста"  # Заменить на фактический шаблон

            keyboard = [
                [InlineKeyboardButton("Подтвердить", callback_data=f'confirm_{channel_name}_{user_id}')],
                [InlineKeyboardButton("Отклонить", callback_data=f'decline_{channel_name}_{user_id}')]
            ]
            reply_markup = InlineKeyboardMarkup(keyboard)

            try:
                await application.bot.send_message(
                    chat_id=owner_id,
                    text=f'Получен запрос на взаимный пост от канала @{channel_name}. Шаблон поста:\n\n{post_template}\n\nПодтвердите или отклоните запрос.',
                    reply_markup=reply_markup
                )
                await query.message.reply_text(f'Шаблон поста для канала @{channel_name} отправлен владельцу канала. Ожидайте подтверждения.')
            except Exception as e:
                logger.error(f'Ошибка при отправке сообщения владельцу канала: {e}')
                await query.message.reply_text('Не удалось отправить запрос владельцу канала.')
        else:
            await query.message.reply_text(f'Не удалось найти владельца канала @{channel_name}.')

        await query.message.delete()

# Функция для обработки подтверждения или отклонения запроса
async def handle_response(update: Update, context: ContextTypes.DEFAULT_TYPE):
    query = update.callback_query
    data = query.data
    user_id = query.from_user.id

    if data.startswith('confirm_'):
        _, channel_name, requesting_user_id = data.split('_')
        requesting_user_id = int(requesting_user_id)

        update_post_request_status(requesting_user_id, channel_name, 'confirmed')

        await query.message.reply_text(f'Запрос на взаимный пост для канала @{channel_name} подтвержден.')
        try:
            await application.bot.send_message(
                chat_id=requesting_user_id,
                text=f'Ваш запрос на взаимный пост в канале @{channel_name} был подтвержден владельцем канала.'
            )
        except Exception as e:
            logger.error(f'Ошибка при отправке сообщения пользователю: {e}')
    
    elif data.startswith('decline_'):
        _, channel_name, requesting_user_id = data.split('_')
        requesting_user_id = int(requesting_user_id)

        update_post_request_status(requesting_user_id, channel_name, 'declined')

        await query.message.reply_text(f'Запрос на взаимный пост для канала @{channel_name} отклонен.')
        try:
            await application.bot.send_message(
                chat_id=requesting_user_id,
                text=f'Ваш запрос на взаимный пост в канале @{channel_name} был отклонен владельцем канала.'
            )
        except Exception as e:
            logger.error(f'Ошибка при отправке сообщения пользователю: {e}')

    await query.message.delete()

# Функция для получения шаблона поста
async def receive_post_template(update: Update, context: ContextTypes.DEFAULT_TYPE):
    user_id = update.message.from_user.id
    text = update.message.text

    if user_id in post_requests:
        channel_name = post_requests[user_id]
        add_post_request(user_id, channel_name, text, 'pending')
        await update.message.reply_text(f'Шаблон поста для канала @{channel_name} отправлен владельцу канала. Ожидайте подтверждения.')
        del post_requests[user_id]
    else:
        await update.message.reply_text('Нет активного запроса.')

# Функция для обработки запросов
async def process_requests(update: Update, context: ContextTypes.DEFAULT_TYPE):
    user_id = update.message.from_user.id
    pending_requests = get_pending_requests(user_id)

    if not pending_requests:
        await update.message.reply_text('У вас нет активных запросов на взаимные посты.')
        return

    response = 'Ваши активные запросы:\n'
    for req in pending_requests:
        response += f'Канал @{req["to_channel"]} - Статус: {req["status"]}\n'
    
    await update.message.reply_text(response)

# Основная функция
def main():
    import database
    database.initialize_db()

    global application
    application = Application.builder().token(TOKEN).build()

    # Обработчики команд и кнопок
    application.add_handler(CommandHandler('start', start))
    application.add_handler(CommandHandler('stats', stats))
    application.add_handler(CommandHandler('addchannel', add_channel))
    application.add_handler(CommandHandler('removechannel', remove_channel))
    application.add_handler(CommandHandler('createpost', create_post))
    application.add_handler(CommandHandler('requests', process_requests))
    application.add_handler(CallbackQueryHandler(handle_response))
    application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, receive_post_template))

    # Запуск бота
    application.run_polling()

if __name__ == '__main__':
    main()

File: bot.js
require("dotenv").config();
const TelegramBot = require("node-telegram-bot-api");
const express = require("express");
const bodyParser = require("body-parser");
const db = require("./database"); // Убедитесь, что этот путь правильный

const token = process.env.TELEGRAM_BOT_TOKEN;
const bot = new TelegramBot(token, { polling: true });
const ADMIN_USER_ID = parseInt(process.env.ADMIN_USER_ID);
const postRequests = {}; // Хранение запросов на взаимные посты

// Функция для проверки прав на отправку сообщений
const checkIfBotCanPostMessages = async (channelName) => {
  const chatId = `@${channelName}`;
  console.log(`Checking permissions for bot in channel ${chatId}`);

  try {
    const chat = await bot.getChat(chatId);
    console.log(`Chat details:`, chat);

    const botInfo = await bot.getMe();
    console.log(`Bot ID: ${botInfo.id}`);

    const chatMember = await bot.getChatMember(chatId, botInfo.id);
    console.log(`Chat member status: ${chatMember.status}`);

    if (
      chatMember.status === "administrator" ||
      chatMember.status === "creator"
    ) {
      console.log(`Bot is an admin in ${chatId}`);
      return true;
    } else {
      console.log(`Bot is not an admin in ${chatId}`);
      return false;
    }
  } catch (error) {
    console.error(
      `Error getting chat member information for ${chatId}:`,
      error.message
    );
    return false;
  }
};

// Команды бота
bot.onText(/\/start/, (msg) => {
  bot.sendMessage(
    msg.chat.id,
    "Привет! Я бот для взаимного обмена постами между каналами."
  );
});

bot.onText(/\/stats/, (msg) => {
  const userId = msg.from.id;
  const username = msg.from.username;

  db.getUserChannels(userId, (err, userChannels) => {
    if (err) {
      console.error(`Error getting user channels from database: ${err}`);
      bot.sendMessage(msg.chat.id, "Произошла ошибка при получении данных.");
      return;
    }

    if (!userChannels || userChannels.length === 0) {
      bot.sendMessage(
        msg.chat.id,
        "У вас нет привязанных каналов. Используйте /addchannel @channel_name для добавления канала."
      );
      return;
    }

    let response = `Личный кабинет для @${username}:\nПривязанные каналы:\n`;
    userChannels.forEach((channel, index) => {
      response += `${index + 1}. @${channel}\n`;
    });

    response += "\nДобавить канал: /addchannel @channel_name\n";
    response += "Удалить канал: /removechannel @channel_name\n";
    response += "Создать запрос на взаимный пост: /createpost\n";

    bot.sendMessage(msg.chat.id, response);
  });
});
bot.onText(/\/addchannel (.+)/, async (msg, match) => {
  const userId = msg.from.id;
  const channelName = match[1].replace("@", "");

  try {
    console.log(
      `Received request to add channel: @${channelName} for user ${userId}`
    );

    const canPostMessages = await checkIfBotCanPostMessages(channelName);

    if (canPostMessages) {
      // Добавляем канал в базу данных
      db.addUserChannel(userId, channelName, (err) => {
        if (err) {
          console.error(`Error adding channel to database: ${err}`);
          bot.sendMessage(
            msg.chat.id,
            "Ошибка при добавлении канала в базу данных."
          );
          return;
        }
        bot.sendMessage(
          msg.chat.id,
          `Канал @${channelName} привязан к вашему аккаунту.`
        );
      });
    } else {
      bot.sendMessage(
        msg.chat.id,
        `Пожалуйста, убедитесь, что я имею права администратора в канале @${channelName}.`
      );
    }
  } catch (error) {
    console.error("Error in /addchannel command:", error);
    bot.sendMessage(
      msg.chat.id,
      `Ошибка при добавлении канала: ${error.message}`
    );
  }
});

bot.onText(/\/removechannel (.+)/, (msg, match) => {
  const userId = msg.from.id;
  const channelName = match[1].replace("@", "");

  db.removeUserChannel(userId, channelName, (err) => {
    if (err) {
      bot.sendMessage(msg.chat.id, "Ошибка при удалении канала.");
      return;
    }
    bot.sendMessage(
      msg.chat.id,
      `Канал @${channelName} отвязан от вашего аккаунта.`
    );
  });
});

bot.onText(/\/createpost/, (msg) => {
  const userId = msg.from.id;

  // Получаем все привязанные каналы
  db.getAllChannels((err, allChannels) => {
    if (err) {
      console.error(`Error getting all channels from database: ${err}`);
      bot.sendMessage(msg.chat.id, "Ошибка при получении всех каналов.");
      return;
    }

    if (!allChannels || allChannels.length === 0) {
      bot.sendMessage(
        msg.chat.id,
        "Нет доступных каналов для создания запроса. Добавьте каналы с помощью /addchannel @channel_name."
      );
      return;
    }

    // Формируем клавиатуру для выбора канала
    const keyboard = allChannels.map((channel) => [
      { text: `@${channel}`, callback_data: `select_channel_${channel}` },
    ]);

    const opts = {
      reply_markup: {
        inline_keyboard: keyboard,
      },
    };

    bot.sendMessage(msg.chat.id, "Выберите канал для взаимного поста:", opts);
  });
});
bot.on("callback_query", async (query) => {
  const chatId = query.message.chat.id;
  const data = query.data;
  const userId = query.from.id;

  if (data.startsWith("select_channel_")) {
    const channelName = data.substring("select_channel_".length);
    postRequests[userId] = { channelName, postTemplate: null };
    bot.sendMessage(
      chatId,
      `Вы выбрали канал @${channelName}. Пожалуйста, отправьте шаблон поста.`
    );
    bot.deleteMessage(chatId, query.message.message_id);
  } else if (data.startsWith("confirm_") || data.startsWith("decline_")) {
    const [action, channelName, requesterId] = data.split("_", 3);
    if (action === "confirm") {
      // Обработка подтверждения
      const postTemplate = postRequests[requesterId]?.postTemplate;
      if (postTemplate) {
        const responseMessage = `${postTemplate}`;
        bot.sendMessage(`@${channelName}`, responseMessage);
        bot.sendMessage(
          requesterId,
          `Ваш запрос на взаимный пост был принят и опубликован в @${channelName}.`
        );
        db.updatePostRequest(
          requesterId,
          channelName,
          postTemplate,
          "confirmed",
          (err) => {
            if (err) {
              console.error("Ошибка при обновлении запроса:", err);
            }
          }
        );
      }
    } else {
      bot.sendMessage(
        requesterId,
        `Ваш запрос на взаимный пост от канала @${channelName} был отклонен.`
      );
    }
    bot.deleteMessage(chatId, query.message.message_id);
    delete postRequests[requesterId];
  }
});

bot.on("message", async (msg) => {
  const chatId = msg.chat.id;
  const text = msg.text;

  if (postRequests[chatId]) {
    const { channelName } = postRequests[chatId];
    postRequests[chatId].postTemplate = text;

    try {
      // Получаем имя канала отправителя
      const senderChannelName = await getChannelNameByUserId(chatId);

      // Получаем ID владельца канала
      const ownerId = await getChannelOwnerId(channelName);

      // Создаем клавиатуру для подтверждения или отклонения запроса
      const keyboard = [
        [
          {
            text: "Подтвердить",
            callback_data: `confirm_${channelName}_${chatId}`,
          },
          {
            text: "Отклонить",
            callback_data: `decline_${channelName}_${chatId}`,
          },
        ],
      ];
      const opts = { reply_markup: { inline_keyboard: keyboard } };

      // Отправляем сообщение владельцу канала для подтверждения запроса
      bot.sendMessage(
        ownerId,
        `Получен запрос на взаимный пост от канала @${senderChannelName}. Шаблон поста:\n\n${text}\n\nПодтвердите или отклоните запрос.`,
        opts
      );

      // Сохраняем запрос в базу данных
      db.addPostRequest(chatId, channelName, text, "pending", (err) => {
        if (err) {
          console.error("Error adding post request:", err);
        }
      });
    } catch (error) {
      console.error("Error in /createpost command:", error);
      bot.sendMessage(
        chatId,
        `Ошибка при создании запроса на пост: ${error.message}`
      );
    }
  }
});

async function getChannelOwnerId(channelName) {
  return new Promise((resolve, reject) => {
    // Здесь предполагается, что у вас есть функция, которая возвращает ID владельца канала по его имени
    // Возможно, вам нужно будет реализовать эту функцию в зависимости от вашей структуры данных
    db.getChannelOwnerId(channelName, (err, ownerId) => {
      if (err) {
        console.error("Ошибка при получении ID владельца канала:", err);
        reject("Неизвестный владелец канала");
      }
      resolve(ownerId);
    });
  });
}



// Запуск сервера
const app = express();
app.use(bodyParser.json());

app.get("/channels/:userId", (req, res) => {
  const userId = req.params.userId;
  db.getUserChannels(userId, (err, userChannels) => {
    if (err) {
      console.error("Error getting user channels:", err);
      res.status(500).send("Ошибка при получении данных.");
      return;
    }
    res.json(userChannels);
  });
});

app.post("/removeChannel", (req, res) => {
  const { userId, channel } = req.body;
  db.removeUserChannel(userId, channel, (err) => {
    if (err) {
      res.status(500).send("Ошибка при удалении канала.");
      return;
    }
    res.send("Канал удален");
  });
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
});

// Запуск бота
bot.on("polling_error", console.log);

console.log("Bot and server are running...");
